src/python/pants/backend/go/subsystems/golang.py

```diff
@@ -3,56 +3,147 @@
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.option.subsystem import Subsystem
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import OrderedSet
 
+logger = logging.getLogger(__name__)
 
-class GolangSubsystem(TemplatedExternalTool):
+
+class GolangSubsystem(Subsystem):
     options_scope = "golang"
-    name = "golang"
-    help = "Official golang distribution."
-
-    default_version = "1.16.5"
-    default_known_versions = [
-        "1.16.5|macos_arm64 |7b1bed9b63d69f1caa14a8d6911fbd743e8c37e21ed4e5b5afdbbaa80d070059|125731583",
-        "1.16.5|macos_x86_64|be761716d5bfc958a5367440f68ba6563509da2f539ad1e1864bd42fe553f277|130223787",
-        "1.16.5|linux_x86_64|b12c23023b68de22f74c0524f10b753e7b08b1504cb7e417eccebdd3fae49061|129049763",
-    ]
-    default_url_template = "https://golang.org/dl/go{version}.{platform}.tar.gz"
-    default_url_platform_mapping = {
-        "macos_arm64": "darwin-arm64",
-        "macos_x86_64": "darwin-amd64",
-        "linux_x86_64": "linux-amd64",
-    }
-
-    def generate_exe(self, plat: Platform) -> str:
-        return "./bin"
+    help = "Options for Golang support."
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--go-search-paths",
+            type=list,
+            member_type=str,
+            default=["<PATH>"],
+            help=(
+                "A list of paths to search for Go.\n\n"
+                "Specify absolute paths to directories with the `go` binary, e.g. `/usr/bin`. "
+                "Earlier entries will be searched first.\n\n"
+                "The special string '<PATH>' will expand to the contents of the PATH env var."
+            ),
+        )
+        # TODO(#13005): Support multiple Go versions in a project?
+        register(
+            "--expected-version",
+            type=str,
+            default="1.17",
+            help=(
+                "The Go version you are using, such as `1.17`.\n\n"
+                "Pants will only use Go distributions from `--go-search-paths` that have the "
+                "expected version, and it will error if none are found. "
+            ),
+        )
+
+    def go_search_paths(self, env: Environment) -> tuple[str, ...]:
+        def iter_path_entries():
+            for entry in self.options.go_search_paths:
+                if entry == "<PATH>":
+                    path = env.get("PATH")
+                    if path:
+                        for path_entry in path.split(os.pathsep):
+                            yield path_entry
+                else:
+                    yield entry
+
+        return tuple(OrderedSet(iter_path_entries()))
+
+    @property
+    def expected_version(self) -> str:
+        return cast(str, self.options.expected_version)
 
 
 @dataclass(frozen=True)
 class GoRoot:
     """Path to the Go installation (the `GOROOT`)."""
 
     path: str
-    digest: Digest
```

Comment:
We can probably keep `digest` and just set it to `EMPTY_DIGEST`. This would prevent having to update all of the call sites to work with and without a system-installed Go tool. I see Pants supporting both downloading Go and system-installed `go` since some users may want to rely on Go installed into the environment (e.g., a Docker image).

Reply:
> I see Pants supporting both downloading Go and system-installed go since some users may want to rely on Go installed into the environment (e.g., a Docker image).

I could see that too, once we support immutable caches. In the meantime, your benchmark shows that Pants-installed Go is too prohibitive so I recommend that we Keep It Simple.